### ISSUE TITLE
htaccess blocks access to the well-know folder

### ISSUE THREAD

**Issue by xmacinfo** [backdrop/backdrop-issues#2451]
The Backdrop .htaccess file blocks access to the .well-known folder, preventing the use of Let's Encrypt to create SSL certificates.

----
~~PR https://github.com/backdrop/backdrop/pull/1713~~
PR https://github.com/backdrop/backdrop/pull/1726

**Comment by jlfranklin** [comment]
If you have access to the Apache configuration, you can put the `.well-known` directory outside of the Backdrop webroot with an Apache config like this:

    Alias /.well-known/acme-challenge/ /var/www/html/.well-known/acme-challenge/
    <Directory "/var/www/html/.well-known/acme-challenge/">
        Options None
        AllowOverride None
        ForceType text/plain
        RedirectMatch 404 "^(?!/\.well-known/acme-challenge/[\w-]{43}$)"
    </Directory>

See https://community.letsencrypt.org/t/apache-multidomain-webroot/10663 for more details.

**Comment by xmacinfo** [comment]
That would be a temporary solution and would not work on shared hosting or on Cpanel.

I had to rename .htaccess to let Let's Encrypt create it's certificate and then renamed .htaccess back to normal, which is still a temporary solution.

With Drupal 7 and Drupal 8 I am using .htaccess successfully:

```
<FilesMatch "\.(engine|inc|info|install|make|module|profile|test|po|sh|.*sql|theme|tpl(\.php)?|xtmpl)(~|\.sw[op]|\.bak|\.orig|\.save)?$|^(\.(?!well-known).*|Entries.*|Repository|Root|Tag|Template|composer\.(json|lock))$|^#.*#$|\.php(~|\.sw[op]|\.bak|\.orig\.save)$">
  Order allow,deny
</FilesMatch>
```

**Comment by Dinornis** [comment]
I was able to replicate this on my hosting service.
Here is the exact wording of the error:

There was a problem processing your request

* Error issuing certificate
* Failed to issue certificate
*The Let's Encrypt HTTP challenge failed: acme error 'urn:acme:error:unauthorized': Invalid response from `http://bdtest.example.com/.well-known/acme-challenge/09MvB24lz1y64IFcecB79AOMI7l66BpjspUdQca8`: "<!DOCTYPE html> <html style="height:100%"> <head><title> 403 Forbidden </title></head> <body style="color: `#444`; margin:0;font:"

**Comment by quicksketch** [comment]
The approach of white-listing `.well-know` sounds good to me, as @xmacinfo suggests in https://github.com/backdrop/backdrop-issues/issues/2451#issuecomment-269940078. Anyone up for making a PR?

**Comment by serundeputy** [comment]
This `PR` https://github.com/backdrop/backdrop/pull/1713 updates the `.htaccess` file outside core; so updating via `drush` or official update instructions will not pick up this change.

@quicksketch Do we have a process for this?  fancy update hook or something?

**Comment by serundeputy** [comment]
@xmacinfo are you up for manually testing this?

**Comment by xmacinfo** [comment]
@serundeputy I will manually test this.

**Comment by quicksketch** [comment]
> @quicksketch Do we have a process for this? fancy update hook or something?

No, if a file like .htaccess is updated, we only leave a note in the release notes that it has changed. Because the .htaccess file is likely to be modified by individual sites and there is no ability to include another file, the documentation is the best we can do. In a situation like this, it's also not terribly important, as we're just enhancing the existing functionality by making it compatible with Let's Encrypt.

**Comment by quicksketch** [comment]
I tested this out but access to the `.well-known` directory still seems to be blocked by this rule later in .htaccess:

```
  # Block access to "hidden" directories whose names begin with a period. This
  # includes directories used by version control systems such as Subversion or
  # Git to store control files. Files whose names begin with a period, as well
  # as the control files used by CVS, are protected by the FilesMatch directive
  # above.
  #
  # NOTE: This only works when mod_rewrite is loaded. Without mod_rewrite, it is
  # not possible to block access to entire directories from .htaccess because
  # <DirectoryMatch> is not allowed here.
  #
  # If you do not have mod_rewrite installed, you should remove these
  # directories from your webroot or otherwise protect them from being
  # downloaded.
  RewriteRule "(^|/)\." - [F]
```

We probably need a similar check in this location.

**Comment by Gormartsen** [comment]
@quicksketch I confirm it. Exactly this line is preventing access to the .well-known .

**Comment by quicksketch** [comment]
I filed a new PR at https://github.com/backdrop/backdrop/pull/1713 that adjusts this later rewrite rule. Oddly it doesn't look like the first rule needs to be changed at all, because it only affects access to individual files, not entire directories like this later one does.

**Comment by jenlampton** [comment]
Well, I wanted to give this some proper testing with letsencrypt, but after about an hour of trying to get Certbot to work on my server, I gave up. Instead I manually created a `.well-known/test.txt` file, and applied the patch. I am able to access that file via apache now, so marking this issue as RTBC.  If anyone who is already using letsencrypt could test, that would be double wonderful :)

**Comment by quicksketch** [comment]
Great, I've merged https://github.com/backdrop/backdrop/pull/1726 into 1.x and 1.5.x.

**Comment by olafgrabienski** [comment]
I had a look at .htaccess of a Drupal 8 site and found that write the rule a bit different.

**Backdrop fix:**

```
RewriteCond %{REQUEST_URI} !^/\.well-known
RewriteRule "(^|/)\." - [F]
```

**D8:**

`RewriteRule "(^|/)\.(?!well-known)" - [F]`

I don't understand the difference, just giving the hint in case one of the coders wants to consider it.

**Comment by olafgrabienski** [comment]
Speaking of Drupal 8's .htaccess: In the *Protect files and directories from prying eyes* section, there is also a well-known related entry: `^(\.(?!well-known).*`

**Comment by quicksketch** [comment]
The difference in the D8/Backdrop fix is just a matter of whether you want to do it all in one line or two. `RewriteCond` lines just put modifiers on the following rule.

I don't think we need the condition inside of the `FilesMatch` directive, as it only affects files and not directories: http://httpd.apache.org/docs/current/mod/core.html#filesmatch Though it would be good to double-check if Drupal had a reason for doing so, or if it was just made based on the assumption it was needed.

**Comment by Gormartsen** [comment]
> if it was just made based on the assumption it was needed.

Any changes to FileMatch does not affect access to .well-known.
I assume that changes were based on assumption as well)

**Comment by Gormartsen** [comment]
I like two line approach. It is much easier to read compare to one line regexp
